hathor/p2p/manager.py

```diff
@@ -560,3 +596,114 @@ def drop_connection_by_peer_id(self, peer_id: str) -> None:
+        try:
+            self._sync_update_cmds()
+        except Exception:
+            self.log.error('_sync_update_cmds failed', exc_info=True)
+
+        try:
+            self._sync_rotate_if_needed()
+        except Exception:
+            self.log.error('_sync_rotate_if_needed failed', exc_info=True)
+
+    def _sync_update_cmds(self) -> None:
+        """Run sync_update commands. The supported commands are:
+
+        1. `always_enable_sync.txt` should contain a list of peer ids that will always have sync enabled.
+           This update skips lines starting with '#'.
+        """
+        cmd_path = self.manager.get_cmd_path()
+        if cmd_path is None:
+            return
+
+        always_sync_path = os.path.join(cmd_path, 'always_enable_sync.txt')
+        if not os.path.isfile(always_sync_path):
+            return
+
+        new = set()
+        with open(always_sync_path, 'r') as fp:
+            for line in fp.readlines():
+                line = line.strip()
+                if not line:
+                    continue
+                if line.startswith('#'):
+                    continue
+                new.add(line)
+
+        old = self.always_enable_sync
+        if new == old:
+            return
+
+        to_enable = new - old
+        to_disable = old - new
+
+        self.log.info('update always_enable_sync', new=new, to_enable=to_enable, to_disable=to_disable)
+
+        for peer_id in new:
+            if peer_id not in self.connected_peers:
+                continue
+            self.connected_peers[peer_id].enable_sync()
+
+        for peer_id in to_disable:
+            if peer_id not in self.connected_peers:
+                continue
+            self.connected_peers[peer_id].disable_sync()
+
+        self.always_enable_sync = new
+
+    def _check_force_sync_rotate(self) -> bool:
+        """Check if sync rotate should forcefully run now."""
+        cmd_path = self.manager.get_cmd_path()
+        if cmd_path is None:
+            return False
+
+        force_rotate_path = os.path.join(cmd_path, 'force_sync_rotate')
+        if not os.path.isfile(force_rotate_path):
+            return False
+
+        self.log.info('force sync rotate detected')
+        os.remove(force_rotate_path)
+        return True
+
+    def _sync_rotate_if_needed(self) -> None:
+        """Rotate peers who we are syncing from."""
+        force = self._check_force_sync_rotate()
+
+        now = self.reactor.seconds()
+        dt = now - self._last_sync_rotate
+        if not force and dt < self.SYNC_UPDATE_INTERVAL:
+            return
+        self._last_sync_rotate = now
+
+        current_enabled: Set[str] = set()
+        for peer_id, conn in self.connected_peers.items():
+            if conn.is_sync_enabled():
+                current_enabled.add(peer_id)
+
+        candidates = list(self.connected_peers.keys())
+        self.rng.shuffle(candidates)
+        selected_peers: Set[str] = set(candidates[:self.MAX_ENABLED_SYNC])
```

Comment:
With the current code we might select here only peers in the `always_enable_sync` list, is this a problem somehow? I don't see any bugs or problems that might happen because of this.

Reply:
I don't think so. The idea is that we must keep the sync enabled for a set of peers. If some of these peers are selected from the candidates, they will be enabled anyway. What do you think?